File: backend/models/summary.py
"""Summary model for LLM-generated event summaries."""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from sqlalchemy import CheckConstraint, DateTime, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from backend.models.camera import Base


class SummaryType(str, Enum):
    """Types of summaries that can be generated.

    Attributes:
        HOURLY: Summary covering the past 60 minutes
        DAILY: Summary covering since midnight today
    """

    HOURLY = "hourly"
    DAILY = "daily"


class Summary(Base):
    """LLM-generated narrative summary of security events.

    Summaries are generated every 5 minutes by a background job and provide
    high-level descriptions of high/critical events for the dashboard.

    The summary contains a natural language narrative (2-4 sentences) describing
    notable security activity in the time window, or a reassuring "all clear"
    message when no high-priority events occurred.

    Attributes:
        id: Auto-increment primary key
        summary_type: Type of summary ('hourly' or 'daily')
        content: LLM-generated narrative text
        event_count: Number of high/critical events included
        event_ids: Array of event IDs that were summarized (nullable)
        window_start: Start of the time window
        window_end: End of the time window
        generated_at: When the LLM produced this summary
        created_at: Row creation timestamp (auto-set by database)
    """

    __tablename__ = "summaries"
    __table_args__ = (
        CheckConstraint(
            "summary_type IN ('hourly', 'daily')",
            name="summaries_type_check",
        ),
        Index("idx_summaries_type_created", "summary_type", "created_at"),
        Index("idx_summaries_created_at", "created_at"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    summary_type: Mapped[str] = mapped_column(String(10), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    event_count: Mapped[int] = mapped_column(Integer, nullable=False)
    event_ids: Mapped[list[int] | None] = mapped_column(ARRAY(Integer), nullable=True)
    window_start: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    window_end: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    generated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    def __repr__(self) -> str:
        return f"<Summary(id={self.id}, type={self.summary_type}, events={self.event_count})>"

File: frontend/src/types/summary.ts
/**
 * Summary Types
 *
 * Types for event summary feature.
 *
 * Summaries are LLM-generated narrative descriptions of high/critical
 * security events, generated every 5 minutes by a background job.
 *
 * @see backend/api/schemas/summary.py - Backend Pydantic schemas
 * @see NEM-2894
 */

// ============================================================================
// Enums
// ============================================================================

/**
 * Type of summary (matches backend SummaryType enum).
 */
export type SummaryType = 'hourly' | 'daily';

// ============================================================================
// Response Types
// ============================================================================

/**
 * A single summary response from the API.
 */
export interface Summary {
  /** Unique identifier */
  id: number;

  /** LLM-generated narrative content */
  content: string;

  /** Number of high/critical events included in this summary */
  eventCount: number;

  /** Start of the time window (ISO 8601 string) */
  windowStart: string;

  /** End of the time window (ISO 8601 string) */
  windowEnd: string;

  /** When this summary was generated (ISO 8601 string) */
  generatedAt: string;
}

/**
 * Response from GET /api/summaries/latest.
 */
export interface SummariesLatestResponse {
  /** Latest hourly summary (past 60 minutes), null if none exists */
  hourly: Summary | null;

  /** Latest daily summary (since midnight), null if none exists */
  daily: Summary | null;
}

// ============================================================================
// WebSocket Types
// ============================================================================

/**
 * WebSocket message for summary updates.
 */
export interface SummaryUpdateMessage {
  type: 'summary_update';
  data: SummariesLatestResponse;
}

// ============================================================================
// Component Props Types
// ============================================================================

/**
 * Props for the SummaryCard component.
 */
export interface SummaryCardProps {
  /** Type of summary to display */
  type: SummaryType;

  /** The summary data, or null if loading/unavailable */
  summary: Summary | null;

  /** Whether the summary is currently loading */
  isLoading?: boolean;

  /** Error message if fetch failed */
  error?: string | null;
}

// ============================================================================
// Hook Types
// ============================================================================

/**
 * Return type for useSummaries hook.
 */
export interface UseSummariesResult {
  /** Latest hourly summary */
  hourly: Summary | null;

  /** Latest daily summary */
  daily: Summary | null;

  /** Whether data is being fetched */
  isLoading: boolean;

  /** Error if fetch failed */
  error: Error | null;

  /** Manually trigger a refetch */
  refetch: () => Promise<void>;
}

// ============================================================================
// Type Guards
// ============================================================================

/**
 * Type guard for Summary objects.
 *
 * @example
 * ```ts
 * if (isSummary(data)) {
 *   // data is Summary
 *   console.log(data.content);
 * }
 * ```
 */
export function isSummary(obj: unknown): obj is Summary {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    'id' in obj &&
    'content' in obj &&
    'eventCount' in obj &&
    'windowStart' in obj &&
    'windowEnd' in obj &&
    'generatedAt' in obj
  );
}

/**
 * Type guard for SummaryUpdateMessage WebSocket messages.
 *
 * @example
 * ```ts
 * if (isSummaryUpdateMessage(msg)) {
 *   // msg is SummaryUpdateMessage
 *   console.log(msg.data.hourly);
 * }
 * ```
 */
export function isSummaryUpdateMessage(msg: unknown): msg is SummaryUpdateMessage {
  return (
    typeof msg === 'object' &&
    msg !== null &&
    'type' in msg &&
    (msg as { type: string }).type === 'summary_update' &&
    'data' in msg
  );
}
